src/commands/cmd_string.cc

```diff
@@ -620,6 +621,201 @@ class CommandCAD : public Commander {
+  }
+
+  Status Execute(Server *srv, Connection *conn, std::string *output) override {
+    redis::String string_db(srv->storage, conn->GetNamespace());
+
+    std::string a;
+    std::string b;
+    auto s1 = string_db.Get(args_[1], &a);
+    auto s2 = string_db.Get(args_[2], &b);
+
+    if (!s1.ok() && !s1.IsNotFound()) {
+      return {Status::RedisExecErr, s1.ToString()};
+    }
+    if (!s2.ok() && !s2.IsNotFound()) {
+      return {Status::RedisExecErr, s2.ToString()};
+    }
+    if (s1.IsNotFound()) a = "";
+    if (s2.IsNotFound()) b = "";
+
+    // Detect string truncation or later overflows.
+    if (a.length() >= UINT32_MAX - 1 || b.length() >= UINT32_MAX - 1) {
+      return {Status::RedisExecErr, "String too long for LCS"};
+    }
+
+    // Compute the LCS using the vanilla dynamic programming technique of
+    // building a table of LCS(x, y) substrings.
+    auto alen = static_cast<uint32_t>(a.length());
+    auto blen = static_cast<uint32_t>(b.length());
+
+    // Allocate the LCS table.
+    uint64_t dp_size = (alen + 1) * (blen + 1);
+    uint64_t bulk_size = dp_size * sizeof(uint32_t);
+    if (bulk_size >= SIZE_MAX || bulk_size / dp_size != sizeof(uint32_t)) {
+      return {Status::RedisExecErr, "Insufficient memory, failed allocating transient memory for LCS"};
+    }
+    if (bulk_size > PROTO_BULK_MAX_SIZE) {
+      return {Status::RedisExecErr, "Insufficient memory, transient memory for LCS exceeds proto-max-bulk-len"};
+    }
+    std::vector<uint32_t> dp(dp_size, 0);
+    auto lcs = [&](const uint32_t i, const uint32_t j) -> uint32_t & { return dp[i * (blen + 1) + j]; };
+
+    // Start building the LCS table.
+    for (uint32_t i = 1; i <= alen; i++) {
+      for (uint32_t j = 1; j <= blen; j++) {
+        if (a[i - 1] == b[j - 1]) {
+          // The len LCS (and the LCS itself) of two
+          // sequences with the same final character, is the
+          // LCS of the two sequences without the last char
+          // plus that last char.
+          lcs(i, j) = lcs(i - 1, j - 1) + 1;
+        } else {
+          // If the last character is different, take the longest
+          // between the LCS of the first string and the second
+          // minus the last char, and the reverse.
+          lcs(i, j) = std::max(lcs(i - 1, j), lcs(i, j - 1));
+        }
+      }
+    }
+
+    // Store the actual LCS string if needed.
+    std::string result;
+    uint32_t idx = lcs(alen, blen);
+
+    // Allocate when we need to compute the actual LCS string.
+    bool compute_lcs = get_idx_ || !get_len_;
+    if (compute_lcs) result.resize(idx);
+
+    // Build a array if we have to emit the matched ranges.
+    std::string matches;
+    uint32_t matches_len = 0;
+
+    uint32_t i = alen;
+    uint32_t j = blen;
+    uint32_t arange_start = alen;  // alen signals that values are not set.
+    uint32_t arange_end = 0;
+    uint32_t brange_start = 0;
+    uint32_t brange_end = 0;
+    while (compute_lcs && i > 0 && j > 0) {
```

Comment:
Check `compute_lcs` in loop is so weird, isn't it?

Reply:
Ok, I have modified it.